rust/kernel/fs.rs

```diff
@@ -0,0 +1,1262 @@
+            tmpfile: None,
+            get_acl: None,
+            set_acl: None,
+            fileattr_set: None,
+            fileattr_get: None,
+            get_offset_ctx: None,
+        };
+
+        const FILE_ADDRESS_SPACE_OPERATIONS: bindings::address_space_operations =
+            bindings::address_space_operations {
+                writepage: None,
+                read_folio: Some(Self::read_folio_callback),
+                writepages: None,
+                dirty_folio: None,
+                readahead: None,
+                write_begin: None,
+                write_end: None,
+                bmap: None,
+                invalidate_folio: None,
+                release_folio: None,
+                free_folio: None,
+                direct_IO: None,
+                migrate_folio: None,
+                launder_folio: None,
+                is_partially_uptodate: None,
+                is_dirty_writeback: None,
+                error_remove_page: None,
+                swap_activate: None,
+                swap_deactivate: None,
+                swap_rw: None,
+            };
+
+        extern "C" fn read_folio_callback(
+            _file: *mut bindings::file,
+            folio: *mut bindings::folio,
+        ) -> i32 {
+            from_result(|| {
+                // SAFETY: All pointers are valid and stable.
+                let inode = unsafe {
+                    &*(*(*folio)
+                        .__bindgen_anon_1
+                        .page
+                        .__bindgen_anon_1
+                        .__bindgen_anon_1
+                        .mapping)
+                        .host
+                        .cast::<INode<T>>()
+                };
+
+                // SAFETY: The C contract guarantees that the folio is valid and locked, with
+                // ownership of the lock transferred to the callee (this function). The folio is
+                // also guaranteed not to outlive this function.
+                T::read_folio(inode, unsafe { LockedFolio::from_raw(folio) })?;
+                Ok(0)
+            })
+        }
+    }
+
+    /// Kernel module that exposes a single read-only file system implemented by `T`.
+    #[pin_data]
+    pub struct Module<T: Type + ?Sized> {
+        #[pin]
+        fs_reg: Registration,
+        _p: PhantomData<T>,
+    }
+
+    impl<T: Type + ?Sized + Sync + Send> crate::InPlaceModule for Module<T> {
+        type Init = impl PinInit<Self, Error>;
+        fn init(module: &'static ThisModule) -> Self::Init {
+            try_pin_init!(Self {
+                fs_reg <- Registration::new::<T>(module),
+                _p: PhantomData,
+            })
+        }
+    }
+
+    /// Declares a kernel module that exposes a single file system.
+    ///
+    /// The `type` argument must be a type which implements the [`Type`] trait. Also accepts various
+    /// forms of kernel metadata.
+    ///
+    /// # Examples
+    ///
+    /// ```ignore
```

Comment:
The test is ignored, but there are hidden lines below. Is the intention to make those work later?

Reply:
They work, I tested them manually.

The reason this is ignored is because kunit doesn't handle examples that include modules: we talked about this some time back, I may even have opened an issue (not sure though).

Anyway, I get the following error without `ignore`:

```
  CC      kernel/module/main.o
error: non-statement macro in statement position: core::arch::global_asm
   --> rust/doctests_kernel_generated.rs:162:1
    |                                                     
162 | / kernel::module_ro_fs! {  
163 | |     type: MyFs,     
164 | |     name: "myfs",                                 
165 | |     author: "Rust for Linux Contributors",
166 | |     description: "My Rust fs",
167 | |     license: "GPL",                               
168 | | }                  
    | |_^                                                                                                            
    |                                                     
    = note: this error originates in the macro `$crate::macros::module` which comes from the expansion of the macro `
kernel::module_ro_fs` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I would be _delighted_ if this is fixed and I can get this example tested automatically.

Reply:
Ah, I forgot about that, thanks! Yeah, I think we discussed it.

A workaround is wrapping it in a `mod { ... }`, which "works", but we end up with built-in kernel modules that run their `init()` like any other, outside the KUnit run.

Perhaps we should have special support for these (i.e. examples that define a kernel module), or perhaps a perform only syntax check on them (not supported either yet).

Reply:
Yeah, I was going to say that actually running the module init wouldn't be ideal but it may be desirable in the sense that we do run the sample; I guess the problem is that we're not in a controlled environment where we can detect failures.

In any case, coming back to the original point about the hidden lines: I wanted to avoid distracting readers with the details of the trait implementation. But it may be helpful... What do you think? 

(I would be ok with making them visible.)

Reply:
> I guess the problem is that we're not in a controlled environment where we can detect failures.

Yeah, exactly. "Luckily", having an `assert!` currently inside a `mod` would not compile anyway.

> In any case, coming back to the original point about the hidden lines: I wanted to avoid distracting readers with the details of the trait implementation. But it may be helpful... What do you think?
>
> (I would be ok with making them visible.)

I think it would be helpful to show the lines, especially because they are forced to implement the trait.

Another argument one may use is that macro documentation pages are simply the docs, i.e. there is no long list of methods, `impl`s, etc. like for types, so having a full example looks better and gives more context for readers looking around (e.g. those reading the list of macros, rather than something from `fs` in particular).